# Code submitted for Cognistx test by Siddhartha Shrivastava
## siddhartha.shri31@gmail.com
## I have used basic approach as time was limited (tf-idf over LR)
## exploration can be done on 1) Word Lemmatizer 2) Stemming 3) LSTM


In [78]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [4]:
train_1 = pd.read_csv('/Users/owner/Downloads/Dataset-3/Train.csv')
test_1 = pd.read_csv('/Users/owner/Downloads/Dataset-3/Test.csv')

In [5]:
train_1.head()

ID                                            comment     date  \
0  uid_590555  Well, let's be honest here, they don't actuall...  2015-04   
1  uid_671762  Well, I didn't need evidence to believe in com...  2016-12   
2  uid_519689              Who does an "official promo" in 360p?  2013-11   
3  uid_788362                           Grotto koth was the best  2015-09   
4  uid_299252                                   Neal's back baby  2015-11   

   down                                     parent_comment  score  top  \
0     0  They should shut the fuck up and let the commu...      2    2   
1    -1  You need evidence to kill people? I thought we...      6   -1   
2     0                    2014 BMW S1000R: Official Promo      3    3   
3     0  Not really that memorable lol if you want memo...      2    2   
4     0                      James Neal hit on Zach Parise     -5   -5   

              topic            user  label  
0       starcitizen  Combat_Wombatz      0  
1  EnoughCommieSpam        starkadd      1  
2       motorcycles         phybere      0  
3        hcfactions          m0xyMC      1  
4            hockey       Somuch101      1

In [6]:
test_1

ID                                            comment     date  \
0      uid_764784  Meanwhile if I pick my boxer up to move him to...  2016-01   
1       uid_67552  Thats what you get for using an unfair advanta...  2015-03   
2      uid_240490  only to be later faced with thousands of dolla...  2010-09   
3       uid_56568  What a thoughtful rebuttal to a well articulat...  2015-12   
4      uid_875860  Yes, major life experiences that demand a grea...  2014-04   
5      uid_808361                       Zero Tolerance at it's best.  2014-10   
6      uid_737105  So you're saying there is still a chance he's ...  2014-09   
7      uid_300028  Because the smaller population states tend to ...  2016-12   
8      uid_851812                             Book learnin i'd guess  2014-03   
9      uid_664649                                               Wat?  2014-01   
10     uid_868017                                  You dropped this?  2015-12   
11      uid_66257                   One lifts, the other cuts clean.  2016-08   
12      uid_66093  How can there possibly be any one 'right' path...  2015-08   
13     uid_593687    Kendall Wright, Justin Hunter, Nate Washington.  2014-10   
14     uid_548795  She cured autism, yes... and for that we owe h...  2016-08   
15     uid_481542                           Hey, that's not Arizona!  2016-05   
16    uid_1010607  That's probably because they are made of marbl...  2016-09   
17     uid_136975             Its a garbage land of snow and sadness  2016-03   
18     uid_230676  Language is far far more offensive than blood ...  2013-08   
19     uid_122978  Zeira from fairy tail zero being a fragment of...  2015-11   
20     uid_583877                                See you there, bro.  2015-07   
21      uid_92063  As someone who doesn't know much about the sub...  2016-06   
22     uid_220399               What battle does Charron do that in?  2013-03   
23      uid_68254  Be proud you don't have a thigh gap because pe...  2014-09   
24     uid_684011                    Plot twist : he's being drugged  2015-06   
25     uid_632384  but then Antonio Brown and Big Ben would have ...  2016-01   
26     uid_433873     It's what music was like before you were born.  2016-07   
27    uid_1001432  But but...Yamcha won because they threw tomato...  2016-09   
28     uid_594981  And yet, you've researched your post so well a...  2010-09   
29     uid_176531  Sure, he's won several Champions Leagues, Seri...  2013-08   
...           ...                                                ...      ...   
7970   uid_895026                            Dahmer did far worse...  2014-07   
7971   uid_580497                                               Sexy  2015-11   
7972   uid_653139                                       Screen Rant.  2016-09   
7973   uid_563562     Hey man, lay off the free market or something.  2015-05   
7974   uid_505242  There are amoebas with more vertebrae than Donald  2016-09   
7975    uid_72241  I think Jim Carrey made a movie about it, but ...  2013-08   
7976   uid_869908                            This is literalty epic.  2015-02   
7977   uid_510322  But but but he's OP so if you don't kill him t...  2016-04   
7978   uid_966353                          Embrace your awkwardness.  2013-03   
7979   uid_516622                         Aw hell no this is awesome  2015-11   
7980   uid_144551                  Yep Arnold, you've done it again!  2015-09   
7981   uid_142287                         there's always whimsyshire  2015-01   
7982   uid_302888  wow so there is a east coast MOD, god fucking ...  2015-10   
7983   uid_388396  Como siempre, solo quieren hablar y exponer su...  2016-07   
7984   uid_472371  To be fair, hematite isn't a jewel, its an ore...  2016-03   
7985   uid_180190  What's so maddening about names like "Samsung ...  2015-01   
7986   uid_770512  Note to self: get a pet duck, name it Dick and...  2016-05   
7987   uid_787829  Seeing all those Japanese b

# Preliminary Checks for both Train and Test dataset

In [7]:
train_1.isnull().sum()

ID                0
comment           0
date              0
down              0
parent_comment    0
score             0
top               0
topic             0
user              0
label             0
dtype: int64

In [11]:
train_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 10 columns):
ID                15000 non-null object
comment           15000 non-null object
date              15000 non-null object
down              15000 non-null int64
parent_comment    15000 non-null object
score             15000 non-null int64
top               15000 non-null int64
topic             15000 non-null object
user              15000 non-null object
label             15000 non-null int64
dtypes: int64(4), object(6)
memory usage: 1.1+ MB


In [8]:
test_1.isnull().sum()

ID                0
comment           0
date              0
down              0
parent_comment    0
score             0
top               0
topic             0
user              0
dtype: int64

In [12]:
test_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 9 columns):
ID                8000 non-null object
comment           8000 non-null object
date              8000 non-null object
down              8000 non-null int64
parent_comment    8000 non-null object
score             8000 non-null int64
top               8000 non-null int64
topic             8000 non-null object
user              8000 non-null object
dtypes: int64(3), object(6)
memory usage: 562.6+ KB


In [10]:
print(train_1.shape)
print(test_1.shape)

(15000, 10)
(8000, 9)


##  Seems like No NaNs

In [13]:
train_1

ID                                            comment  \
0       uid_590555  Well, let's be honest here, they don't actuall...   
1       uid_671762  Well, I didn't need evidence to believe in com...   
2       uid_519689              Who does an "official promo" in 360p?   
3       uid_788362                           Grotto koth was the best   
4       uid_299252                                   Neal's back baby   
5        uid_29449  Orange is the New Black House of Cards Hemlock...   
6       uid_528005             Pff everybody knows science is sexist.   
7       uid_850110  because he called out r/mma's golden boy sage ...   
8       uid_840172                What if he's not into dudes though?   
9       uid_288113  probably should've followed your original comm...   
10      uid_171248              That's gonna look great later in life   
11      uid_310752  Nonsense, the bookstore has competitive prices...   
12      uid_853461  at least asus is known for putting quality ram...   
13      uid_556411                     Don't equate stance with rice.   
14      uid_489762  Since when did federal laws start applying to ...   
15      uid_299999  Take a fistful of shrooms and go watch The Hol...   
16       uid_71795                I'd date me, I'd date me so hard...   
17      uid_135024  I thought this sub was about balance, not fun ...   
18      uid_288999  oh yeah well i have a doctorate in c++ and a m...   
19      uid_950946  Wow, I just glanced through your history of ra...   
20      uid_659836                        What took you guys so long?   
21      uid_903725  They could have just put ' I suck john oliver ...   
22      uid_636504                                       Got you, fam   
23      uid_925603  It gets worse, one of the top pro players was ...   
24      uid_381619                                   He's a sexy man.   
25       uid_72802  Yeah but Reddit isn't a billion dollar busines...   
26      uid_542479                      Wow so no credit to Bargnani?   
27      uid_199388                             its tee time doctoh xd   
28      uid_841449                                        absolutely!   
29      uid_617401  Lol, well it's more a trademark issue, but yea...   
...            ...                                                ...   
14970   uid_176610                                               Wtf?   
14971     uid_8219               Literally all of that is changeable.   
14972   uid_397311                                     Can I have job   
14973   uid_201969        So how's that whole Free Tibet thing going?   
14974   uid_882379  But they get free games, can't blame the cost ...   
14975   uid_480579                It's a lot of fucking carbon, mate.   
14976   uid_626238         Does watching Manziel play count as abuse?   
14977    uid_82517  These poor Syrian "refugees", seriously we all...   
14978    uid_88728  Temp chains + chilled ground + tar zombies map...   
14979     uid_4615  After the 08-09 Cup finals as well as the qual...   
14980    uid_92707                                      you dropped a   
14981   uid_926696        HE CROSSED THIS N*GGA INTO THE SHADOW REALM   
14982    uid_97644     Too bad there aren't things called lane-swaps.   
14983   uid_671646  It's okay as long as the thing it kills has a ...   
14984    uid_95764         Too bad reading the rules is not required.   
14985   uid_818374                           I would be a shitty fan.   
14986   uid_177774  Obviously it's because anyone who has over 100...   
14987    uid_56115                      Exactly what a bot would say.   
14988   uid_895549                                              Maybe   
14989   uid_155156  Log into your dashboard and look under driver ...   
14990   uid_338598  If she skipped a meal a day she'd gain 50 poun...   
14991   uid_254764                                          Psychotic   
14992   uid_574041                  Hehe, your text flair says "duty"   
1

In [197]:
# we can see they are equally divided, makes it a bit tricky
train_1['label'].value_counts()

1    7527
0    7473
Name: label, dtype: int64

In [88]:
from sklearn.model_selection import train_test_split

# Here we split our data for the train and test sets
train_df, test_df, train_y, test_y = train_test_split(train_1, train_1['label'], test_size=0.20, random_state=17)

# EDA

In [203]:
# EDA
EDA_train = train_1[['comment', 'user', 'topic','date', 'score','parent_comment','label']].copy()

In [204]:
EDA_train.head()

comment            user  \
0  Well, let's be honest here, they don't actuall...  Combat_Wombatz   
1  Well, I didn't need evidence to believe in com...        starkadd   
2              Who does an "official promo" in 360p?         phybere   
3                           Grotto koth was the best          m0xyMC   
4                                   Neal's back baby       Somuch101   

              topic     date  score  \
0       starcitizen  2015-04      2   
1  EnoughCommieSpam  2016-12      6   
2       motorcycles  2013-11      3   
3        hcfactions  2015-09      2   
4            hockey  2015-11     -5   

                                      parent_comment  label  
0  They should shut the fuck up and let the commu...      0  
1  You need evidence to kill people? I thought we...      1  
2                    2014 BMW S1000R: Official Promo      0  
3  Not really that memorable lol if you want memo...      1  
4                      James Neal hit on Zach Parise      1

## Topic wise Frequency check 

In [206]:
filtered = EDA_train.groupby(['topic']).filter(lambda x: x['comment'].count()>100)
filtered.groupby(['topic']).agg({
    'comment':'count',
    'label': 'mean'}).sort_values(by='label', ascending=False)

comment     label
topic                             
worldnews            399  0.671679
TumblrInAction       125  0.632000
politics             572  0.624126
news                 254  0.602362
soccer               117  0.589744
leagueoflegends      331  0.589124
nfl                  221  0.579186
nba                  187  0.577540
todayilearned        189  0.555556
pcmasterrace         267  0.554307
hockey               130  0.538462
GlobalOffensive      185  0.529730
videos               185  0.529730
AdviceAnimals        199  0.517588
CFB                  103  0.504854
WTF                  144  0.500000
gaming               178  0.500000
pics                 260  0.469231
funny                286  0.468531
The_Donald           156  0.416667
DotA2                101  0.415842
AskReddit            942  0.393843

## Some Users are notorious for Sarcastic comments

In [207]:
freq_users = EDA_train.groupby(['user']).filter(lambda x: x['comment'].count()>2)
freq_users.groupby(['user']).agg({
    'comment':'count',
    'label': 'mean'}).sort_values(by='label', ascending=False)

comment     label
user                                  
zerofuxstillhungry         3  1.000000
Mordredbas                 3  1.000000
IrishJoe                   3  1.000000
Insane_Artist              3  1.000000
IceRollMenu2               3  1.000000
mattreyu                   3  1.000000
Roddy0608                  3  1.000000
Gurchimo                   3  1.000000
FNAFfan69                  3  1.000000
Gamer_152                  5  0.800000
berniebrah                 4  0.750000
Satancake                  4  0.750000
ryeshoes                   4  0.750000
TodaysIllusion             4  0.750000
reaper527                  4  0.750000
mindlessrabble             4  0.750000
jdepps113                  4  0.750000
AudibleNod                 4  0.750000
dafirnz                    3  0.666667
lovelybone93               3  0.666667
Pepe_leprawn               3  0.666667
staiano                    3  0.666667
PrayForMojo_               3  0.666667
RedLegionnaire             3  0.666667
Renegade_Meister           3  0.666667
metans                     3  0.666667
meaniereddit               3  0.666667
Republinuts                3  0.666667
lawson_dlaw                3  0.666667
david-me                   3  0.666667
...                      ...       ...
Halo_likes_me              3  0.333333
smjpilot                   3  0.333333
FlyingBallWithWings        3  0.333333
richardleosimones          3  0.333333
Erzherzog                  3  0.333333
reddixmadix                3  0.333333
VeteranKamikaze            3  0.333333
Voyddd                     3  0.333333
WhimsyUU                   3  0.333333
DoctorEwser                3  0.333333
Natanael_L                 3  0.333333
Dirtydeedsinc              3  0.333333
murderhuman                6  0.333333
-lumpinator-               4  0.250000
VanderpumpDrools           4  0.250000
SWIMsfriend                4  0.250000
ParentalGuidance-13        4  0.250000
InFearn0                   4  0.250000
Zadok_The_Priest           5  0.200000
purebishop                 5  0.200000
nhlfan40                   3  0.000000
XA36                       3  0.000000
Wemorg                     3  0.000000
TheThinMan34               3  0.000000
ACE_C0ND0R                 3  0.000000
Springtrapattacks          4  0.000000
PianoRainMelody            3  0.000000
Llort2                     3  0.000000
Chaoslab                   3  0.000000
Strazdas1                  3  0.000000

[114 rows x 2 columns]

## well users individual comments are low, so we can focus on other factors

In [208]:
# Here we add the new field -  comment_len
EDA_train['comment_length'] = EDA_train['comment'].apply(len)

In [209]:
EDA_train.head()

comment            user  \
0  Well, let's be honest here, they don't actuall...  Combat_Wombatz   
1  Well, I didn't need evidence to believe in com...        starkadd   
2              Who does an "official promo" in 360p?         phybere   
3                           Grotto koth was the best          m0xyMC   
4                                   Neal's back baby       Somuch101   

              topic     date  score  \
0       starcitizen  2015-04      2   
1  EnoughCommieSpam  2016-12      6   
2       motorcycles  2013-11      3   
3        hcfactions  2015-09      2   
4            hockey  2015-11     -5   

                                      parent_comment  label  comment_length  
0  They should shut the fuck up and let the commu...      0             127  
1  You need evidence to kill people? I thought we...      1              53  
2                    2014 BMW S1000R: Official Promo      0              37  
3  Not really that memorable lol if you want memo...      1              24  
4                      James Neal hit on Zach Parise      1              16

## Now it's time to explore the text features. First let's look at the most popular words for sarcastic comments.We will throw away most common words, like 'a', 'the', 'and' etc. and will concider unigramms , bigramms and Trigram separately

In [211]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

In [212]:
vectorizer_1 = CountVectorizer(stop_words='english', ngram_range=(1, 1))
vectorizer_2 = CountVectorizer(stop_words='english', ngram_range=(2, 2))
vectorizer_3 = CountVectorizer(stop_words='english', ngram_range=(3, 3))

In [98]:
def freq_words(vectorizer, data):
    X = vectorizer.fit_transform(data)
    freqs = zip(vectorizer.get_feature_names(), np.asarray(X.sum(axis=0)).ravel())
    return sorted(freqs, key = lambda x: x[1], reverse=True)[:10]

In [215]:
# OUTPUT ----> [(sarcastic comments,freq ), (non-sarcastic comments, freq)]
list_com = [freq_words(vectorizer_1, EDA_train[EDA_train['label']==1]['comment']),
     freq_words(vectorizer_1, EDA_train[EDA_train['label']==0]['comment'])]
list(map(list, zip(*list_com)))

[[('just', 471), ('just', 468)],
 [('yeah', 457), ('like', 451)],
 [('like', 372), ('don', 300)],
 [('people', 323), ('people', 224)],
 [('don', 294), ('think', 200)],
 [('right', 216), ('good', 176)],
 [('know', 203), ('know', 161)],
 [('good', 186), ('really', 153)],
 [('sure', 183), ('make', 149)],
 [('really', 163), ('time', 149)]]

In [100]:
#lets try bigram

In [216]:
# OUTPUT ----> [(sarcastic comments,freq ), (non-sarcastic comments, freq)]

list_com = [freq_words(vectorizer_2, EDA_train[EDA_train['label']==1]['comment']),
     freq_words(vectorizer_2, EDA_train[EDA_train['label']==0]['comment'])]
list(map(list, zip(*list_com)))

[[('don know', 38), ('f5 f5', 79)],
 [('just like', 33), ('starbucks starbucks', 53)],
 [('sounds like', 29), ('looks like', 30)],
 [('good thing', 21), ('don think', 29)],
 [('oh yeah', 20), ('pretty sure', 22)],
 [('white people', 20), ('don know', 19)],
 [('looks like', 19), ('sounds like', 16)],
 [('didn know', 16), ('don want', 15)],
 [('mikekiller mikekiller', 16), ('feel like', 15)],
 [('black people', 15), ('look like', 14)]]

In [217]:
#lets try trigram
# OUTPUT ----> [(sarcastic comments,freq ), (non-sarcastic comments, freq)]

list_com = [freq_words(vectorizer_3, EDA_train[EDA_train['label']==1]['comment']),
     freq_words(vectorizer_3, EDA_train[EDA_train['label']==0]['comment'])]
list(map(list, zip(*list_com)))

[[('mikekiller mikekiller mikekiller', 15), ('f5 f5 f5', 78)],
 [('howard jimmy howard', 11), ('starbucks starbucks starbucks', 52)],
 [('jimmy howard jimmy', 11), ('doo doo doo', 8)],
 [('broken shore spoilers', 3), ('naught naught naught', 5)],
 [('check privilege shitlord', 3), ('just make sure', 4)],
 [('comment shitting shit', 3), ('kisses kisses kisses', 3)],
 [('like year old', 3), ('people don need', 3)],
 [('makes perfect sense', 3), ('11 59 pm', 2)],
 [('oh looks like', 3), ('12 years old', 2)],
 [('pics didn happen', 3), ('18 year old', 2)]]

# Above exploration didn't provide any helpful division as we can see. Both the 'Sarcastic' and 'Non- Sarcastic' Unigram, Bigram and Trigram has almost same words.
# So lets try simple approach of Tf-idf over Logistic regression

In [218]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
import string
#from nltk.stem import PorterStemmer, WordNetLemmatizer

In [219]:
log_reg = LogisticRegression(random_state=17)

In [220]:
X = train_1['comment']
y = train_1['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=28)


In [221]:
from sklearn.pipeline import Pipeline


In [222]:
Model = Pipeline([('tfidf',TfidfVectorizer(min_df=2, ngram_range=(1,3),use_idf=True)),
                    ('logit',LogisticRegression(solver='lbfgs', max_iter=3000))])
Model.fit(X_train,y_train)

Pipeline(memory=None,
     steps=[('tfidf', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.float64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=2,
        ngram_range=(1, 3), norm='l2', preprocessor=None, smooth_idf=True,...enalty='l2', random_state=None, solver='lbfgs',
          tol=0.0001, verbose=0, warm_start=False))])

In [223]:
from sklearn.metrics import confusion_matrix, accuracy_score


In [224]:
predictions = Model.predict(X_test)
print('Accuracy Score: {:.2%}'.format(accuracy_score(y_test,predictions)),'\n')

Accuracy Score: 66.33% 



## dividing training and testing in 70:30 sample size gave 65.04% accuracy
## dividing training and testing in 80:20 sample size gave 66.33% accuracy

In [225]:
s1=f1_score(y_test,predictions,average='weighted')
s1

0.663108187904184


# F1 score is ~66%

In [167]:
predictions

array([1, 0, 0, ..., 0, 1, 0])

In [144]:
X_test.head()

2105          Yes, all I do all day is calculate fugacity.
3714                                 I don't quite follow.
6692     Also: utility bills, dental visits, car mainte...
13479                                I think he was joking
9317                                Must be double dipped.
Name: comment, dtype: object

# now using Model on "testing data"

In [83]:
test_1.head()

ID                                            comment     date  \
0  uid_764784  Meanwhile if I pick my boxer up to move him to...  2016-01   
1   uid_67552  Thats what you get for using an unfair advanta...  2015-03   
2  uid_240490  only to be later faced with thousands of dolla...  2010-09   
3   uid_56568  What a thoughtful rebuttal to a well articulat...  2015-12   
4  uid_875860  Yes, major life experiences that demand a grea...  2014-04   

   down                                     parent_comment  score  top  \
0     0                             Part of the Family Fun      2    2   
1     0  received a warning for hacking while overclock...      1    1   
2     0  Running shoes and race entries. I started runn...      1    1   
3     0  Actually, you should get over your self. There...     -1   -1   
4     0  As a bonus, it was with his ex, months after t...     -5   -5   

           topic                  user  
0           gifs  Cameltoe-Swampdonkey  
1           h1z1            Gothika_47  
2         Frugal                 tells  
3            CFB        YourToothbrush  
4  AdviceAnimals   drunken_trophy_wife

In [188]:
test_predictions = Model.predict(test_1['comment'])

In [189]:
test_1['label'] = test_predictions

# Finally converting in desired .csv format

In [226]:
final_file = test_1[['ID','label']]

In [227]:
final_file

ID  label
0      uid_764784      0
1       uid_67552      0
2      uid_240490      0
3       uid_56568      1
4      uid_875860      1
5      uid_808361      0
6      uid_737105      1
7      uid_300028      1
8      uid_851812      1
9      uid_664649      0
10     uid_868017      1
11      uid_66257      0
12      uid_66093      1
13     uid_593687      0
14     uid_548795      1
15     uid_481542      0
16    uid_1010607      1
17     uid_136975      1
18     uid_230676      0
19     uid_122978      0
20     uid_583877      1
21      uid_92063      0
22     uid_220399      0
23      uid_68254      1
24     uid_684011      1
25     uid_632384      1
26     uid_433873      0
27    uid_1001432      1
28     uid_594981      1
29     uid_176531      0
...           ...    ...
7970   uid_895026      0
7971   uid_580497      0
7972   uid_653139      0
7973   uid_563562      1
7974   uid_505242      0
7975    uid_72241      0
7976   uid_869908      0
7977   uid_510322      1
7978   uid_966353      1
7979   uid_516622      0
7980   uid_144551      0
7981   uid_142287      0
7982   uid_302888      1
7983   uid_388396      0
7984   uid_472371      0
7985   uid_180190      0
7986   uid_770512      1
7987   uid_787829      1
7988   uid_163343      0
7989   uid_354875      1
7990   uid_826429      0
7991   uid_137492      0
7992   uid_982401      1
7993   uid_123364      1
7994   uid_892472      0
7995   uid_893847      0
7996    uid_34003      1
7997   uid_921530      1
7998   uid_440748      0
7999   uid_815387      0

[8000 rows x 2 columns]

In [228]:
final_file.to_csv("final_file.csv",header=['ID', 'label'],index=False)